course_UvA-DL/transformers-and-MH-attention/Transformers_and_MHAttention.py

```diff
@@ -0,0 +1,1491 @@
+# Furthermore, and most importantly, there is so much more to the Transformer architecture.
+# NLP is the domain the Transformer architecture has been originally proposed for and had the greatest impact on, but it also accelerated research in other domains, recently even [Computer Vision](https://arxiv.org/abs/2010.11929).
+# Thus, we focus here on what makes the Transformer and self-attention so powerful in general.
+# In a second notebook, we will look at Vision Transformers, i.e. Transformers for image classification ([link to notebook](https://uvadlc-notebooks.readthedocs.io/en/latest/tutorial_notebooks/tutorial15/Vision_Transformer.html)).
+#
+# Below, we import our standard libraries.
+
+import json
+import math
+import os
+import random
+import urllib.request
+from functools import partial
+# %%
+# Standard libraries
+from urllib.error import HTTPError
+
+import matplotlib
+# Imports for plotting
+import matplotlib.pyplot as plt
+import numpy as np
+import pytorch_lightning as pl
+import seaborn as sns
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+import torch.optim as optim
+import torch.utils.data as data
+import torchvision
+from IPython.display import set_matplotlib_formats
+from matplotlib.colors import to_rgb
+from pytorch_lightning.callbacks import LearningRateMonitor, ModelCheckpoint
+from torchvision import transforms
+from torchvision.datasets import CIFAR100
+from tqdm.notebook import tqdm
+
+plt.set_cmap('cividis')
+# %matplotlib inline
+set_matplotlib_formats('svg', 'pdf')  # For export
+matplotlib.rcParams['lines.linewidth'] = 2.0
+sns.reset_orig()
+
+# tqdm for loading bars
+
+# PyTorch
+
+# Torchvision
+
+# PyTorch Lightning
+
+# Path to the folder where the datasets are/should be downloaded (e.g. CIFAR10)
+DATASET_PATH = os.environ.get('PATH_DATASETS', "data/")
+# Path to the folder where the pretrained models are saved
+CHECKPOINT_PATH = os.environ.get('PATH_CHECKPOINT', "saved_models/Transformers/")
+
+# Setting the seed
+pl.seed_everything(42)
+
+# Ensure that all operations are deterministic on GPU (if used) for reproducibility
+torch.backends.cudnn.determinstic = True
+torch.backends.cudnn.benchmark = False
+
+device = torch.device("cuda:0") if torch.cuda.is_available() else torch.device("cpu")
+print("Device:", device)
+
+# %% [markdown]
+# Two pre-trained models are downloaded below.
+# Make sure to have adjusted your `CHECKPOINT_PATH` before running this code if not already done.
+
+# %%
+# Github URL where saved models are stored for this tutorial
+base_url = "https://raw.githubusercontent.com/phlippe/saved_models/main/tutorial6/"
```

Comment:
@Borda @phlippe do we have an option to save these models somewhere else?

@phlippe Nothing personal but I just don't feel comfortable with storing models for one of the tutorials somewhere we don't have direct access. Hope you understand :)

Reply:
Sure, we can also store them somewhere else. Do you have a suggestion where? I had it so far in a separate repository to reduce the size of the notebook repo.

Reply:
Sure, we can save the models in our s3 buckets.